fix freezing gui performance

trittsv opened the pull request:
When there was a single topic with more then 200+ endpoints the gui was not useable, it was freezing to death.
This PR fixes the performance issues, it now fully uses Signals/Slot (like qt suggest) and a performance fix in the qos check.
The data handling is now more Object-Oriented, resulting in less looping to avoid unneeded cpu time.

- fix freezing gui - tested it with 600+ endpoints on a topic.
- fix performance hungry QoS mismatch check, compare only smallest needed subset.
- fix app crash when user copied something to clipboard.
- fix layout, last endpoint was only half visible.
- fix missing scroll bar in endpoint overview.
- fix gui-model update, use add/remove instead reset model every time.

@eboasson may you can have a look? 😃

eboasson reviewed (commented):
Thanks @trittsv 😀

It looks good to me, except for one trivial typo and a slight worry about the disappearance of a mutex. That worry may well be caused by my very limited knowledge on python threading and Qt behaviour.

I did make a few more comments, but they are really about alternatives that I think might make sense. It is quite possible you considered these already and rejected them, but perhaps there's a detail in them that happens to be of some value. None of those seem to be worth making changes to this PR, but I'll leave that you to decide.

trittsv commented on src/dds_data.py:
```diff
@@ -10,208 +10,333 @@
+        self.endpoint: DcpsEndpoint = endpoint
+        self.entity_type: EntityType = entity_type
+        self.participant = None
+        self.missmatches : Dict[str, List[dds_qos_policy_id]]= {}
```
Thank you for finding the typo, i fixed it.

trittsv commented on src/dds_data.py:
```diff
@@ -10,208 +10,333 @@
+        if endpointKey in self.reader_endpoints:
+            for mimKey in self.reader_endpoints[endpointKey].missmatches:
+                if mimKey in self.writer_endpoints:
+                    if endpointKey in self.writer_endpoints[mimKey].missmatches:
```
yes you are right the `if` in line 65 is may a bit over guardedly. 
In the mismatches of a reader can only be writer ids.
Since we are working on a copy here there should be no need for a lock.

Good id to use the guid and not the generic string, i will do that in a separate PR:

trittsv commented:
@eboasson thank you very much for you review and comments.

- I fixed the typo.
- Regarding the absence of mutex, its not needed anymore.
Qts Signal/Slot mechanism with `Qt::QueuedConnection` is thread safe.
Emitting the signal makes a copy and queues the data which den executed in the thread of the slot.
The different thread now only communicate via signals/slot i removed all direct references.

trittsv commented on src/dds_data.py:
```diff
@@ -10,208 +10,333 @@
+
+            del self.writer_endpoints[endpointKey]
+
+    def link_participant(self, participant: DcpsParticipant):
```
i was not sure if the participant always received before the endpoint, thank you for this valuable information. In the next PRs i will adjust it, it safes a view ifs

trittsv commented on src/dds_data.py:
```diff
@@ -10,208 +10,333 @@
+
+    def check_qos_mismatch(self, data_endpoint: DataEndpoint):
+
+        endpoints_to_check = self.reader_endpoints
```
so you mean it could be totally fine to same topic name `x` with readers and writers of different types?
From my experience with DDS systems that sounds like not a good idea, but of course when dds allows it we should look at it.
I will add an issue for that to look at it later.

trittsv commented on src/dds_data.py:
```diff
@@ -10,208 +10,333 @@
+        return self.entity_type == EntityType.WRITER
+
+    def link_participant(self, participant: DcpsParticipant):
+        if str(self.endpoint.participant_key) == str(participant.key):
```
Thank for that information's, this will be resolved in the following PR mentioned above when the `link_participant` will be removed.

eboasson reviewed (approved):
Thank you for confirming the bit about the mutex (and fixing the typo) 🙏
